zkevm-circuits/src/mpt_circuit/account_leaf.rs

```diff
@@ -0,0 +1,589 @@
+            );
+            config.is_balance_mod = IsEqualGadget::construct(
+                &mut cb.base,
+                config.main_data.proof_type.expr(),
+                MPTProofType::BalanceChanged.expr(),
+            );
+            config.is_storage_mod = IsEqualGadget::construct(
+                &mut cb.base,
+                config.main_data.proof_type.expr(),
+                MPTProofType::StorageChanged.expr(),
+            );
+            config.is_codehash_mod = IsEqualGadget::construct(
+                &mut cb.base,
+                config.main_data.proof_type.expr(),
+                MPTProofType::CodeHashExists.expr(),
+            );
+
+            // Drifted leaf handling
+            config.drifted = DriftedGadget::construct(
+                cb,
+                &config.parent_data,
+                &config.key_data,
+                &key_rlc,
+                &leaf_no_key_rlc,
+                &drifted_bytes,
+                &ctx.r,
+            );
+
+            // Wrong leaf handling
+            config.wrong = WrongGadget::construct(
+                cb,
+                a!(ctx.mpt_table.address_rlc),
+                config.is_non_existing_account_proof.expr(),
+                &config.rlp_key[true.idx()].key_value,
+                &key_rlc[true.idx()],
+                &wrong_bytes,
+                config.is_in_empty_trie[true.idx()].expr(),
+                config.key_data[true.idx()].clone(),
+                &ctx.r,
+            );
+            let is_non_existing_account = not::expr(config.wrong.is_key_equal.clone().expr());
+
+            // Anything following this node is below the account
+            // TODO(Brecht): For non-existing accounts it should be impossible to prove
+            // storage leaves unless it's also a non-existing proof?
+            MainData::store(
+                cb,
+                &ctx.memory[main_memory()],
+                [
+                    config.main_data.proof_type.expr(),
+                    true.expr(),
+                    is_non_existing_account.expr(),
+                    key_rlc[true.idx()].expr(),
+                    config.main_data.root_prev.expr(),
+                    config.main_data.root.expr(),
+                ],
+            );
+
+            ifx! {config.is_account_delete_mod => {
+                // Account delete
+                // We need to make sure there is no leaf when account is deleted. Two possible
+                // cases:
+                // - 1. Account leaf is deleted and there is a nil object in
+                // branch. In this case we have a placeholder leaf.
+                // - 2. Account leaf is deleted from a branch with two leaves, the remaining
+                // leaf moves one level up and replaces the branch. In this case we
+                // have a branch placeholder.
+                // TODO(Brecht): For case 2: just having the parent branch be the placeholder seems not enough
```

Comment:
@Brechtpd The other constraints are covered by `DriftedGadget`, right?

Reply:
Yeah seems ok to me now, not sure if I thought of any other problems